### PR TITLE
Feature verification of image sequences

### LINKED ISSUE #537: Verification of image sequences
A subtask may involve rendering more than one frame. Even if there's only one frame this might not necessarily be the first frame. We need to take it into account and render the right frames during verification.

`TaskToCompute.compute_task_def.extra_data.frames` specifies frame numbers to be rendered.

To tell Blender to render a specific frame use the `-f` option. Don't use `-s`/`-e` - this will render multiple frames into a video file which is not what we want. Each frame should be a separate image.

#### Concent Core
Concent Core should validate `TaskToCompute.compute_task_def.extra_data.frames`
- It's a list
- Each frame number is a non-negative integer

#### Conductor
- Add `frames` parameter to `blender_verification_request` and `blender_verification_order`. It should be the list of frames.
- Add `Frame` model to the `storage` database with the following fields:
    - `blender_subtask_definition`: foreign key to `BlenderSubtaskDefinition`, non-null.
    - `number`: int, non-null.
    - `BlenderSubtaskDefinition` and `number` must be unique together.

#### Verifier
- Immediately after unpacking the result package from provider:
    - Parse file names in the archive to determine which frames are present.
    - Compare the frame numbers present there with the ones received in `blender_verification_order`.
        - If any of the frames listed in the order is not present, report mismatch and end verification.
        - If there are more frames than necessary, log a warning but continue verification.
- Run Blender separately for each frame.
    - Run only one Blender process at a time. Don't start rendering the next frame until the previous one finishes.
    - The number of threads used by Blender for rendering each frame should be a setting: `BLENDER_THREADS`, int, default = 1.
- Upload all rendered frames to the storage server (as separate files)
- Compare all the frame pairs.

### Frame names in the result package
Please try to run Golem with a rendering task containing multiple frames. Look into the archive it produces and see how the frames are named.

### PR DESCRIPTION
**Pull request by kbeker**
Resolves #537

### CODE REVIEW

File: concent_api/conductor/tasks.py [hunk 1]
```diff
@@ -151,3 +156,52 @@ def upload_acknowledged(
...
+
+
+def filter_frames_by_blender_subtask_verification(blender_subtask_definition: BlenderSubtaskDefinition):
+    frames_query = Frame.objects.filter(
+        blender_subtask_definition=blender_subtask_definition,
+    )
+    frames = []
+    for frame in frames_query:
+        frames.append(frame.number)
+    return frames
```
Comment (dybi) [inline concent_api/conductor/tasks.py]: why not `return [frame.number for frame in frames_query]` ? ;)

File: concent_api/core/validation.py [hunk 2]
```diff
@@ -411,3 +412,18 @@ def validate_secure_hash_algorithm(checksum: str):
             "Invalid SHA1 hash.",
             ErrorCode.MESSAGE_FILES_CHECKSUM_INVALID_SHA1_HASH
         )
+
+
+def validate_number_of_frames(frames_list: List[int]):
+    if not isinstance(frames_list, list):
+        raise FrameNumbersValidationError(
+            'TaskToCompute must contain list of frames_list.',
```
Comment (dybi) [inline concent_api/core/validation.py]: typo; should be `"list of frames"` ;)

File: concent_api/core/validation.py [hunk 3]
```diff
@@ -411,3 +412,18 @@ def validate_secure_hash_algorithm(checksum: str):
             "Invalid SHA1 hash.",
             ErrorCode.MESSAGE_FILES_CHECKSUM_INVALID_SHA1_HASH
         )
+
+
+def validate_number_of_frames(frames_list: List[int]):
```
Comment (dybi) [inline concent_api/core/validation.py]: please remember about unit tests ;)

File: concent_api/verifier/tasks.py [hunk 4]
```diff
@@ -104,29 +110,68 @@ def blender_verification_order(
...
-    render_image(frame_number, output_format, scene_file, subtask_id)
-
-    delete_source_files(package_paths_to_downloaded_archive_names[source_package_path])
+    result_files_list = get_files_list_from_archive(
+        generate_verifier_storage_file_path(package_paths_to_downloaded_archive_names[result_package_path])
+    )
 
-    blender_output_file_name = generate_full_blender_output_file_name(scene_file, frame_number, output_format.lower())
-    try_to_upload_blender_output_file(blender_output_file_name, output_format, subtask_id)
+    if len(frames) > len(result_files_list):
```
Comment (dybi) [inline concent_api/verifier/tasks.py]: this should be extracted to a separate function ;)

File: concent_api/verifier/tasks.py [hunk 5]
```diff
@@ -104,29 +110,68 @@ def blender_verification_order(
...
+    )
+
+    ssim_list = []
+    for (result_file, blender_output_file_name) in parsed_files_to_compare.values():
+        image_1, image_2 = load_images(
+            blender_output_file_name,
+            result_file,
+            subtask_id
+        )
+        if not are_image_sizes_and_color_channels_equal(image_1, image_2):
```
Comment (dybi) [inline concent_api/verifier/tasks.py]: please extract it to separate function

File: concent_api/verifier/utils.py [hunk 6]
```diff
@@ -354,3 +357,28 @@ def download_archives_from_storage(file_transfer_token, subtask_id, package_path
                 ErrorCode.VERIFIER_FILE_DOWNLOAD_FAILED,
                 subtask_id=subtask_id,
             )
+
+
+def parse_result_files_with_frames(frames, result_files_list, output_format):
+    parsed_files_to_compare = {}
+    for frame_number in frames:
+        for result_file_name in result_files_list:
+            if re.search(f'_{frame_number:>04}.{output_format.lower()}', result_file_name) is not None:
```
Comment (dybi) [inline concent_api/verifier/utils.py]: please add "$" at the end of the regex

File: concent_api/verifier/utils.py [hunk 7]
```diff
@@ -354,3 +357,28 @@ def download_archives_from_storage(file_transfer_token, subtask_id, package_path
...
+                if not parsed_files_to_compare.get(output_format):
```
Comment (dybi) [inline concent_api/verifier/utils.py]: please rewrite to: `if not result_file_name in parsed_files_to_compare.values()`

File: concent_api/verifier/utils.py [hunk 8]
```diff
@@ -354,3 +357,28 @@ def download_archives_from_storage(file_transfer_token, subtask_id, package_path
                 ErrorCode.VERIFIER_FILE_DOWNLOAD_FAILED,
                 subtask_id=subtask_id,
             )
+
+
+def parse_result_files_with_frames(frames, result_files_list, output_format):
+    parsed_files_to_compare = {}
```
Comment (dybi) [inline concent_api/verifier/utils.py]: please rename:
1. `frames_to_result_files_map` or
2. `result_files_to_frames_map`

File: concent_api/verifier/utils.py [hunk 9]
```diff
@@ -354,3 +357,28 @@ def download_archives_from_storage(file_transfer_token, subtask_id, package_path
...
+    for frame_number in frames:
+        for result_file_name in result_files_list:
+            if re.search(f'_{frame_number:>04}.{output_format.lower()}', result_file_name) is not None:
+                if not parsed_files_to_compare.get(output_format):
+                    parsed_files_to_compare.update({frame_number: [generate_verifier_storage_file_path(result_file_name)]})
+                    result_files_list.remove(result_file_name)
```
Comment (dybi) [inline concent_api/verifier/utils.py]: this line should be removed; please do not delete elements of a list, you are currently iterating over ;)

File: concent_api/verifier/utils.py [hunk 10]
```diff
@@ -354,3 +357,28 @@ def download_archives_from_storage(file_transfer_token, subtask_id, package_path
...
             )
+
+
+def parse_result_files_with_frames(frames, result_files_list, output_format):
+    parsed_files_to_compare = {}
+    for frame_number in frames:
+        for result_file_name in result_files_list:
+            if re.search(f'_{frame_number:>04}.{output_format.lower()}', result_file_name) is not None:
+                if not parsed_files_to_compare.get(output_format):
+                    parsed_files_to_compare.update({frame_number: [generate_verifier_storage_file_path(result_file_name)]})
```
Comment (dybi) [inline concent_api/verifier/utils.py]: this can be straightforward: `parsed_files_to_compare[frame_number] = [generate_verifier_storage_file_path(result_file_name)]`

File: concent_api/verifier/utils.py [hunk 11]
```diff
@@ -354,3 +357,28 @@ def download_archives_from_storage(file_transfer_token, subtask_id, package_path
...
+                    result_files_list.remove(result_file_name)
+    return parsed_files_to_compare
+
+
+def render_images_by_frames(parsed_files_to_compare, frames, output_format, scene_file, subtask_id):
+    blender_output_file_name_list = []
+    for frame_number in frames:
+        render_image(frame_number, output_format, scene_file, subtask_id)
+        blender_output_file_name_list.append(generate_full_blender_output_file_name(scene_file, frame_number, output_format.lower()))
+        parsed_files_to_compare[frame_number].append(generate_full_blender_output_file_name(scene_file, frame_number, output_format.lower()))
```
Comment (dybi) [inline concent_api/verifier/utils.py]: please extract `generate_full_blender_output_file_name(scene_file, frame_number, output_format.lower())` to a variable and reuse

File: concent_api/conductor/models.py [hunk 12]
```diff
@@ -83,3 +84,15 @@ class UploadReport(Model):
 
     # Indicates when conductor has been notified about the upload.
     created_at      = DateTimeField(default=timezone.now)
+
+
+class Frame(Model):
+
+    blender_subtask_definition = ForeignKey(BlenderSubtaskDefinition, related_name='frames', null=False)
```
Comment (rwrzesien) [inline concent_api/conductor/models.py]: Isn't `null=False` by default ?

File: concent_api/conductor/models.py [hunk 13]
```diff
@@ -83,3 +84,15 @@ class UploadReport(Model):
...
+
+    number = PositiveIntegerField(null=False)
```
Comment (rwrzesien) [inline concent_api/conductor/models.py]: Same here, isn't `null=False` by default ?

File: concent_api/conductor/models.py [hunk 14]
```diff
@@ -83,3 +84,15 @@ class UploadReport(Model):
 
     # Indicates when conductor has been notified about the upload.
     created_at      = DateTimeField(default=timezone.now)
+
+
+class Frame(Model):
+
```
Comment (rwrzesien) [inline concent_api/conductor/models.py]: You can add a comment what this model represents.

File: concent_api/conductor/tasks.py [hunk 15]
```diff
@@ -1,4 +1,5 @@
 import logging
+from typing import List
```
Comment (rwrzesien) [inline concent_api/conductor/tasks.py]: Alpha.

File: concent_api/conductor/tasks.py [hunk 16]
```diff
@@ -28,37 +30,37 @@ def blender_verification_request(
...
         logger,
         f'Blender verification request starts. SUBTASK_ID: {subtask_id}',
         f'Source_package_path {source_package_path}',
         f'Result_package_path: {result_package_path}',
         f'Output_format: {output_format}',
-        f'Scene_file: {scene_file}'
+        f'Scene_file: {scene_file}',
+        f'Frames: {frames}',
     )
+    assert isinstance(frames, list)
```
Comment (rwrzesien) [inline concent_api/conductor/tasks.py]: You may check if it is a list of ints.

File: concent_api/conductor/tasks.py [hunk 17]
```diff
@@ -151,3 +156,52 @@ def upload_acknowledged(
...
+        store_frame = Frame(
+            blender_subtask_definition=blender_subtask_definition,
+            number=frame,
+        )
+        store_frame.full_clean()
+        store_frame.save()
+
+
+def filter_frames_by_blender_subtask_verification(blender_subtask_definition: BlenderSubtaskDefinition):
+    frames_query = Frame.objects.filter(
```
Comment (rwrzesien) [inline concent_api/conductor/tasks.py]: Actually this can be achieved by single query, using `.fields('number')` and `.values_list()`

File: concent_api/conductor/tasks.py (hunk 1 above)
Comment (kbeker) [reply concent_api/conductor/tasks.py]: Done

File: concent_api/core/validation.py (hunk 2 above)
Comment (kbeker) [reply concent_api/core/validation.py]: Done

File: concent_api/core/validation.py (hunk 3 above)
Comment (kbeker) [reply concent_api/core/validation.py]: Done

File: concent_api/verifier/tasks.py (hunk 4 above)
Comment (kbeker) [reply concent_api/verifier/tasks.py]: Done

File: concent_api/verifier/tasks.py (hunk 5 above)
Comment (kbeker) [reply concent_api/verifier/tasks.py]: Done

File: concent_api/verifier/utils.py (hunk 6 above)
Comment (kbeker) [reply concent_api/verifier/utils.py]: Done

File: concent_api/verifier/utils.py (hunk 8 above)
Comment (kbeker) [reply concent_api/verifier/utils.py]: Done

File: concent_api/verifier/utils.py (hunk 9 above)
Comment (kbeker) [reply concent_api/verifier/utils.py]: Done

File: concent_api/verifier/utils.py (hunk 7 above)
Comment (kbeker) [reply concent_api/verifier/utils.py]: Done

File: concent_api/verifier/utils.py (hunk 11 above)
Comment (kbeker) [reply concent_api/verifier/utils.py]: Done

File: concent_api/verifier/utils.py (hunk 10 above)
Comment (kbeker) [reply concent_api/verifier/utils.py]: Done

File: concent_api/conductor/tasks.py (hunk 15 above)
Comment (kbeker) [reply concent_api/conductor/tasks.py]: Done

File: concent_api/conductor/tasks.py (hunk 16 above)
Comment (kbeker) [reply concent_api/conductor/tasks.py]: Done

File: concent_api/conductor/tasks.py (hunk 17 above)
Comment (kbeker) [reply concent_api/conductor/tasks.py]: Done

File: concent_api/verifier/utils.py [hunk 18]
```diff
@@ -354,3 +354,61 @@ def download_archives_from_storage(file_transfer_token, subtask_id, package_path
...
+        parsed_files_to_compare[frame_number].append(blender_out_file_name)
+    return (blender_output_file_name_list, parsed_files_to_compare)
+
+
+def upload_blender_output_file(frames, blender_output_file_name_list, output_format, subtask_id):
+    for (frame_number, blender_output_file_name) in zip(frames, blender_output_file_name_list):
+        try_to_upload_blender_output_file(blender_output_file_name, output_format, subtask_id, frame_number)
+
+
+def validate_frames_number_and_result_files_list(frames, result_files_list, subtask_id):
```
Comment (dybi) [inline concent_api/verifier/utils.py]: maybe ~ `ensure_enough_result_files_provided`

File: concent_api/verifier/utils.py [hunk 19]
```diff
@@ -354,3 +354,61 @@ def download_archives_from_storage(file_transfer_token, subtask_id, package_path
...
+    if len(frames) > len(result_files_list):
+        raise VerificationMismatch(subtask_id=subtask_id)
+
+    elif len(frames) < len(result_files_list):
+        logger.warning(f'There is more result files than frames to render')
+
+
+def validate_frames_number_and_parsed_files_to_compare(frames, parsed_files_to_compare, subtask_id):
```
Comment (dybi) [inline concent_api/verifier/utils.py]: maybe sth like  `ensure_frames_have_related_files_to_compare`

File: concent_api/core/exceptions.py [hunk 20]
```diff
@@ -19,3 +19,7 @@ class HashingAlgorithmError(ConcentBaseException):
 
 class GolemMessageValidationError(ConcentBaseException):
     pass
+
+
+class FrameNumbersValidationError(ConcentBaseException):
```
[the remark this answers is not in the document]
Comment (kbeker) [reply concent_api/core/exceptions.py]: Done

File: concent_api/core/validation.py [hunk 21]
```diff
@@ -411,3 +412,24 @@ def validate_secure_hash_algorithm(checksum: str):
...
+    if not isinstance(frames_list, list) and frames_list is not None:
```
[the remark this answers is not in the document]
Comment (kbeker) [reply concent_api/core/validation.py]: Done

File: concent_api/core/validation.py [hunk 22]
```diff
@@ -411,3 +412,24 @@ def validate_secure_hash_algorithm(checksum: str):
             "Invalid SHA1 hash.",
             ErrorCode.MESSAGE_FILES_CHECKSUM_INVALID_SHA1_HASH
         )
+
+
+def validate_number_of_frames(frames_list: List[int]):
```
[the remark this answers is not in the document]
Comment (kbeker) [reply concent_api/core/validation.py]: Done

File: concent_api/core/tests/test_unit_validation.py [hunk 23]
```diff
@@ -233,3 +235,36 @@ def test_that_if_different_public_keys_are_given_method_should_return_false(self
...
+
+class TestFramesListValidaation(TestCase):
+
+    def test_that_if_frames_list_is_a_list_with_ints_method_should_pass(self):
+        try:
+            validate_number_of_frames([1, 2])
+        except Exception:  # pylint: disable=broad-except
+            self.fail()
+
+    def test_that_if_frames_is_not_a_list_with_int_method_should_raise_exception(self):
```
[the remark this answers is not in the document]
Comment (kbeker) [reply concent_api/core/tests/test_unit_validation.py]: Done

File: concent_api/core/tests/test_unit_validation.py [hunk 24]
```diff
@@ -233,3 +235,36 @@ def test_that_if_different_public_keys_are_given_method_should_return_false(self
         subtask_results_accepted_list = self.create_subtask_results_accepted_list(self.task_to_compute_1, self.task_to_compute_2)
         result = are_keys_and_addresses_unique_in_message_subtask_results_accepted(subtask_results_accepted_list)
         self.assertFalse(result)
+
+
+class TestFramesListValidaation(TestCase):
+
+    def test_that_if_frames_list_is_a_list_with_ints_method_should_pass(self):
```
[the remark this answers is not in the document]
Comment (kbeker) [reply concent_api/core/tests/test_unit_validation.py]: Done

File: concent_api/verifier/utils.py (hunk 18 above)
Comment (kbeker) [reply concent_api/verifier/utils.py]: Done

File: concent_api/verifier/utils.py (hunk 19 above)
Comment (kbeker) [reply concent_api/verifier/utils.py]: Done

File: concent_api/core/message_handlers.py [hunk 25]
```diff
@@ -1305,17 +1306,14 @@ def handle_send_subtask_results_verify(
     compute_task_def = task_to_compute.compute_task_def
     requestor_public_key = hex_to_bytes_convert(task_to_compute.requestor_public_key)
     provider_public_key = hex_to_bytes_convert(task_to_compute.provider_public_key)
-
     validate_golem_message_subtask_results_rejected(subtask_results_rejected)
-    validate_that_golem_messages_are_signed_with_key(
```
Comment (rwrzesien) [inline concent_api/core/message_handlers.py]: Why is this removed ?

File: concent_api/verifier/utils.py [hunk 26]
```diff
@@ -354,3 +354,61 @@ def download_archives_from_storage(file_transfer_token, subtask_id, package_path
                 ErrorCode.VERIFIER_FILE_DOWNLOAD_FAILED,
                 subtask_id=subtask_id,
             )
+
+
+def parse_result_files_with_frames(frames, result_files_list, output_format):
+    frames_to_result_files_map = {}
+    for frame_number in frames:
+        for result_file_name in result_files_list:
+            if re.search(f'_{frame_number:>04}.{output_format.lower()}$', result_file_name) is not None:
```
Comment (rwrzesien) [inline concent_api/verifier/utils.py]: You can combine this into single condition.

File: concent_api/verifier/utils.py [hunk 27]
```diff
@@ -354,3 +354,61 @@ def download_archives_from_storage(file_transfer_token, subtask_id, package_path
...
+                if result_file_name not in frames_to_result_files_map.values():
+                    frames_to_result_files_map[frame_number] = [generate_verifier_storage_file_path(result_file_name)]
+    return frames_to_result_files_map
+
+
+def render_images_by_frames(parsed_files_to_compare, frames, output_format, scene_file, subtask_id):
```
Comment (rwrzesien) [inline concent_api/verifier/utils.py]: You can add typings and a comment to each of the functions below.

File: concent_api/core/message_handlers.py (hunk 25 above)
Comment (kbeker) [reply concent_api/core/message_handlers.py]: O, if I am to be honest I don't know. It might be deleted during some of rebeases, but I'm not sure. Of course restored.

File: concent_api/verifier/utils.py (hunk 27 above)
Comment (kbeker) [reply concent_api/verifier/utils.py]: Done

File: concent_api/verifier/utils.py (hunk 26 above)
Comment (kbeker) [reply concent_api/verifier/utils.py]: Done

File: concent_api/core/message_handlers.py [hunk 28]
```diff
@@ -1316,6 +1316,8 @@ def handle_send_subtask_results_verify(
         report_computed_task,
     )
 
+    validate_frames(compute_task_def['extra_data']['frames'])
```
Comment (rwrzesien) [inline concent_api/core/message_handlers.py]: Shouldn't we validate this in other use cases too ? If yes, it should be in `validate_task_to_compute` function.

File: concent_api/core/validation.py [hunk 29]
```diff
@@ -411,3 +412,24 @@ def validate_secure_hash_algorithm(checksum: str):
             "Invalid SHA1 hash.",
             ErrorCode.MESSAGE_FILES_CHECKSUM_INVALID_SHA1_HASH
         )
+
+
+def validate_frames(frames_list: List[int]):
+    if not isinstance(frames_list, list) and len(frames_list) > 0:
```
Comment (rwrzesien) [inline concent_api/core/validation.py]: If `frames_list` will be for example `1`, this check will fail with `TypeError: object of type 'int' has no len()`. There should be `or` there instead of `and`.

File: concent_api/verifier/tasks.py [hunk 30]
```diff
@@ -13,24 +14,28 @@
...
 from verifier.utils import validate_downloaded_archives
-from .utils import are_image_sizes_and_color_channels_equal
-from .utils import generate_full_blender_output_file_name
+from .utils import get_files_list_from_archive
+from .utils import generate_verifier_storage_file_path
+from .utils import compare_minimum_ssim_with_results
+from .utils import render_images_by_frames
+from .utils import parse_result_files_with_frames
+from .utils import upload_blender_output_file
+from .utils import compare_all_rendered_images_with_user_results_files
```
Comment (rwrzesien) [inline concent_api/verifier/tasks.py]: It would be cool to have this in alphabetic order :)

File: concent_api/verifier/tests/test_utils.py [hunk 31]
```diff
@@ -1,33 +1,200 @@
...
 
+from core.constants import VerificationResult
+from verifier.exceptions import VerificationMismatch
 from verifier.utils import are_image_sizes_and_color_channels_equal
+from verifier.utils import compare_all_rendered_images_with_user_results_files
+from verifier.utils import compare_minimum_ssim_with_results
+from verifier.utils import render_images_by_frames
+from verifier.utils import parse_result_files_with_frames
+from verifier.utils import upload_blender_output_file
+from verifier.utils import ensure_enough_result_files_provided
```
Comment (rwrzesien) [inline concent_api/verifier/tests/test_utils.py]: Same here.

File: concent_api/core/message_handlers.py (hunk 28 above)
Comment (kbeker) [reply concent_api/core/message_handlers.py]: I have asked @cameel https://github.com/golemfactory/concent/issues/537#issuecomment-402443056 and I'm waiting for answer :)

File: concent_api/verifier/tests/test_utils.py [hunk 32]
```diff
@@ -1,33 +1,200 @@
...
+                scene_file=self.scene_file,
+                subtask_id=self.subtask_id,
+                verification_deadline=None
+            )
+            self.assertEqual(self.correct_parsed_all_files, parsed_files_to_compare)
+            self.assertEqual(self.correct_blender_output_file_name_list, blender_output_file_name_list)
+            self.assertEqual(mock_render_image.call_count, 2)
+
+    def test_that_upload_blender_output_file_should_correctly_upload_files(self):
+        with mock.patch('verifier.utils.try_to_upload_blender_output_file', autospec=True) as mock_try_to_upload:
```
Comment (rwrzesien) [inline concent_api/verifier/tests/test_utils.py]: You should add raising exception as `side_effect` in this mock. Currently it is not clear why exception is being raised there.

File: concent_api/verifier/tests/test_utils.py [hunk 33]
```diff
@@ -1,33 +1,200 @@
...
+            )
+
+    def test_that_method_should_accept_frames_and_result_files_when_correct_variable_is_passed(self):
+        try:
+            ensure_enough_result_files_provided(
+                frames=self.frames,
+                result_files_list=self.result_files_list,
+                subtask_id=self.subtask_id,
+            )
+        except Exception:  # pylint: disable=broad-except
```
Comment (rwrzesien) [inline concent_api/verifier/tests/test_utils.py]: Same here

File: concent_api/verifier/tests/test_utils.py [hunk 34]
```diff
@@ -1,33 +1,200 @@
...
+            )
+
+    def test_that_method_should_accept_correct_frames_list_and_parsed_result_files(self):
+        try:
+            ensure_frames_have_related_files_to_compare(
+                frames=self.frames,
+                parsed_files_to_compare=self.correct_parsed_all_files,
+                subtask_id=self.subtask_id,
+            )
+        except Exception:  # pylint: disable=broad-except
```
Comment (rwrzesien) [inline concent_api/verifier/tests/test_utils.py]: Same here.

File: concent_api/core/validation.py (hunk 29 above)
Comment (kbeker) [reply concent_api/core/validation.py]: Done

File: concent_api/core/message_handlers.py (hunk 28 above)
Comment (kbeker) [reply concent_api/core/message_handlers.py]: @cameel already answered that we should check it always so I added to validate_task_to_compute

File: concent_api/core/tests/test_unit_validation.py [hunk 35]
```diff
@@ -233,3 +235,36 @@ def test_that_if_different_public_keys_are_given_method_should_return_false(self
...
         self.assertFalse(result)
+
+
+class TestFramesListValidaation(TestCase):
+
+    def test_that_list_of_ints_is_valid(self):
+        try:
+            validate_frames([1, 2])
+        except Exception:  # pylint: disable=broad-except
+            self.fail()
```
Comment (cameel) [inline concent_api/core/tests/test_unit_validation.py]: Please remove these exception handlers. Pylint is right here.

That's exactly the reason why we have https://github.com/Code-Poets/coding-style/issues/8
All the warnings you're silencing should be listed in a comment for review.

File: concent_api/verifier/tests/test_utils.py (hunk 33 above)
Comment (cameel) [reply concent_api/verifier/tests/test_utils.py]: Pylint is right. This handler should be removed. The test framework will catch the exception anyway.

If you do it like this you silence everything, even stuff like `SyntaxError` and we don't want that.
